File: client/src-tauri/src/commands/contracts/developer_tool_harness.rs
use serde::{Deserialize, Serialize};
use serde_json::Value as JsonValue;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DeveloperToolPackSummaryDto {
    pub pack_id: String,
    pub label: String,
    pub policy_profile: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DeveloperToolCatalogEntryDto {
    pub tool_name: String,
    pub group: String,
    pub description: String,
    pub tags: Vec<String>,
    pub schema_fields: Vec<String>,
    pub examples: Vec<String>,
    pub risk_class: String,
    pub effect_class: String,
    pub runtime_available: bool,
    pub allowed_runtime_agents: Vec<String>,
    pub activation_groups: Vec<String>,
    pub tool_packs: Vec<DeveloperToolPackSummaryDto>,
    pub input_schema: Option<JsonValue>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DeveloperToolCatalogResponseDto {
    pub host_os: String,
    pub host_os_label: String,
    pub skill_tool_enabled: bool,
    pub entries: Vec<DeveloperToolCatalogEntryDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DeveloperToolHarnessCallDto {
    pub tool_name: String,
    pub input: JsonValue,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tool_call_id: Option<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DeveloperToolHarnessRunOptionsDto {
    #[serde(default)]
    pub stop_on_failure: Option<bool>,
    #[serde(default)]
    pub approve_writes: Option<bool>,
    #[serde(default)]
    pub operator_approve_all: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DeveloperToolSyntheticRunRequestDto {
    pub project_id: String,
    #[serde(default)]
    pub agent_session_id: Option<String>,
    pub calls: Vec<DeveloperToolHarnessCallDto>,
    #[serde(default)]
    pub options: Option<DeveloperToolHarnessRunOptionsDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DeveloperToolHarnessCallResultDto {
    pub tool_call_id: String,
    pub tool_name: String,
    pub ok: bool,
    pub summary: String,
    pub output: JsonValue,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DeveloperToolSyntheticRunResponseDto {
    pub run_id: String,
    pub agent_session_id: String,
    pub stopped_early: bool,
    pub had_failure: bool,
    pub results: Vec<DeveloperToolHarnessCallResultDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DeveloperToolDryRunRequestDto {
    pub project_id: String,
    pub tool_name: String,
    pub input: JsonValue,
    #[serde(default)]
    pub tool_call_id: Option<String>,
    #[serde(default)]
    pub operator_approved: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DeveloperToolPolicyDecisionDto {
    pub action: String,
    pub code: String,
    pub explanation: String,
    pub risk_class: String,
    pub project_trust: String,
    pub network_intent: String,
    pub credential_sensitivity: String,
    pub prior_observation_required: bool,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub os_target: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DeveloperToolDryRunResponseDto {
    pub tool_call_id: String,
    pub tool_name: String,
    pub decoded: bool,
    pub policy_decision: DeveloperToolPolicyDecisionDto,
    pub sandbox_decision: JsonValue,
    pub sandbox_denied: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DeveloperToolSequenceRecordDto {
    pub id: String,
    pub name: String,
    pub calls: Vec<DeveloperToolHarnessCallDto>,
    #[serde(default)]
    pub options: Option<DeveloperToolHarnessRunOptionsDto>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DeveloperToolSequenceListResponseDto {
    pub sequences: Vec<DeveloperToolSequenceRecordDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DeveloperToolSequenceUpsertRequestDto {
    #[serde(default)]
    pub id: Option<String>,
    pub name: String,
    pub calls: Vec<DeveloperToolHarnessCallDto>,
    #[serde(default)]
    pub options: Option<DeveloperToolHarnessRunOptionsDto>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DeveloperToolSequenceDeleteRequestDto {
    pub id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DeveloperToolModelRunRequestDto {
    pub project_id: String,
    pub tool_name: String,
    pub prompt: String,
    #[serde(default)]
    pub agent_session_id: Option<String>,
    #[serde(default)]
    pub runtime_agent_id: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DeveloperToolModelRunResponseDto {
    pub run_id: String,
    pub agent_session_id: String,
    pub project_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct DeveloperToolHarnessProjectDto {
    pub project_id: String,
    pub display_name: String,
    pub root_path: String,
}

File: client/src/lib/xero-model/developer-tool-harness.ts
import { z } from 'zod'

export const developerToolPackSummarySchema = z
  .object({
    packId: z.string().trim().min(1),
    label: z.string(),
    policyProfile: z.string(),
  })
  .strict()
export type DeveloperToolPackSummaryDto = z.infer<typeof developerToolPackSummarySchema>

export const developerToolCatalogEntrySchema = z
  .object({
    toolName: z.string().trim().min(1),
    group: z.string(),
    description: z.string(),
    tags: z.array(z.string()),
    schemaFields: z.array(z.string()),
    examples: z.array(z.string()),
    riskClass: z.string(),
    effectClass: z.string(),
    runtimeAvailable: z.boolean(),
    allowedRuntimeAgents: z.array(z.string()),
    activationGroups: z.array(z.string()),
    toolPacks: z.array(developerToolPackSummarySchema),
    inputSchema: z.unknown().nullable().optional(),
    runtimeUnavailableReason: z.string().nullable().optional(),
  })
  .strict()
export type DeveloperToolCatalogEntryDto = z.infer<typeof developerToolCatalogEntrySchema>

export const developerToolCatalogResponseSchema = z
  .object({
    hostOs: z.string(),
    hostOsLabel: z.string(),
    skillToolEnabled: z.boolean(),
    entries: z.array(developerToolCatalogEntrySchema),
  })
  .strict()
export type DeveloperToolCatalogResponseDto = z.infer<typeof developerToolCatalogResponseSchema>

export const developerToolCatalogRequestSchema = z
  .object({
    skillToolEnabled: z.boolean().optional(),
  })
  .strict()
export type DeveloperToolCatalogRequestDto = z.infer<typeof developerToolCatalogRequestSchema>

export const developerToolHarnessCallSchema = z
  .object({
    toolName: z.string().trim().min(1),
    input: z.unknown(),
    toolCallId: z.string().min(1).optional(),
  })
  .strict()
export type DeveloperToolHarnessCallDto = z.infer<typeof developerToolHarnessCallSchema>

export const developerToolHarnessRunOptionsSchema = z
  .object({
    stopOnFailure: z.boolean().optional(),
    approveWrites: z.boolean().optional(),
    operatorApproveAll: z.boolean().optional(),
  })
  .strict()
export type DeveloperToolHarnessRunOptionsDto = z.infer<
  typeof developerToolHarnessRunOptionsSchema
>

export const developerToolSyntheticRunRequestSchema = z
  .object({
    projectId: z.string().trim().min(1),
    agentSessionId: z.string().min(1).optional(),
    calls: z.array(developerToolHarnessCallSchema).min(1),
    options: developerToolHarnessRunOptionsSchema.optional(),
  })
  .strict()
export type DeveloperToolSyntheticRunRequestDto = z.infer<
  typeof developerToolSyntheticRunRequestSchema
>

export const developerToolHarnessCallResultSchema = z
  .object({
    toolCallId: z.string(),
    toolName: z.string(),
    ok: z.boolean(),
    summary: z.string(),
    output: z.unknown(),
  })
  .strict()
export type DeveloperToolHarnessCallResultDto = z.infer<
  typeof developerToolHarnessCallResultSchema
>

export const developerToolSyntheticRunResponseSchema = z
  .object({
    runId: z.string(),
    agentSessionId: z.string(),
    stoppedEarly: z.boolean(),
    hadFailure: z.boolean(),
    results: z.array(developerToolHarnessCallResultSchema),
  })
  .strict()
export type DeveloperToolSyntheticRunResponseDto = z.infer<
  typeof developerToolSyntheticRunResponseSchema
>

export const developerToolDryRunRequestSchema = z
  .object({
    projectId: z.string().trim().min(1),
    toolName: z.string().trim().min(1),
    input: z.unknown(),
    toolCallId: z.string().min(1).optional(),
    operatorApproved: z.boolean().optional(),
  })
  .strict()
export type DeveloperToolDryRunRequestDto = z.infer<typeof developerToolDryRunRequestSchema>

export const developerToolPolicyDecisionSchema = z
  .object({
    action: z.string(),
    code: z.string(),
    explanation: z.string(),
    riskClass: z.string(),
    projectTrust: z.string(),
    networkIntent: z.string(),
    credentialSensitivity: z.string(),
    priorObservationRequired: z.boolean(),
    osTarget: z.string().nullable().optional(),
  })
  .strict()
export type DeveloperToolPolicyDecisionDto = z.infer<typeof developerToolPolicyDecisionSchema>

export const developerToolDryRunResponseSchema = z
  .object({
    toolCallId: z.string(),
    toolName: z.string(),
    decoded: z.boolean(),
    policyDecision: developerToolPolicyDecisionSchema,
    sandboxDecision: z.unknown(),
    sandboxDenied: z.boolean(),
  })
  .strict()
export type DeveloperToolDryRunResponseDto = z.infer<typeof developerToolDryRunResponseSchema>

export const developerToolSequenceRecordSchema = z
  .object({
    id: z.string().min(1),
    name: z.string().min(1),
    calls: z.array(developerToolHarnessCallSchema),
    options: developerToolHarnessRunOptionsSchema.nullable().optional(),
    createdAt: z.string().min(1),
    updatedAt: z.string().min(1),
  })
  .strict()
export type DeveloperToolSequenceRecordDto = z.infer<typeof developerToolSequenceRecordSchema>

export const developerToolSequenceListResponseSchema = z
  .object({
    sequences: z.array(developerToolSequenceRecordSchema),
  })
  .strict()
export type DeveloperToolSequenceListResponseDto = z.infer<
  typeof developerToolSequenceListResponseSchema
>

export const developerToolSequenceUpsertRequestSchema = z
  .object({
    id: z.string().min(1).optional(),
    name: z.string().trim().min(1),
    calls: z.array(developerToolHarnessCallSchema).min(1),
    options: developerToolHarnessRunOptionsSchema.optional(),
  })
  .strict()
export type DeveloperToolSequenceUpsertRequestDto = z.infer<
  typeof developerToolSequenceUpsertRequestSchema
>

export const developerToolSequenceDeleteRequestSchema = z
  .object({
    id: z.string().min(1),
  })
  .strict()
export type DeveloperToolSequenceDeleteRequestDto = z.infer<
  typeof developerToolSequenceDeleteRequestSchema
>

export const developerToolModelRunRequestSchema = z
  .object({
    projectId: z.string().trim().min(1),
    toolName: z.string().trim().min(1),
    prompt: z.string().trim().min(1),
    agentSessionId: z.string().min(1).optional(),
    runtimeAgentId: z.string().min(1).optional(),
  })
  .strict()
export type DeveloperToolModelRunRequestDto = z.infer<typeof developerToolModelRunRequestSchema>

export const developerToolModelRunResponseSchema = z
  .object({
    runId: z.string(),
    agentSessionId: z.string(),
    projectId: z.string(),
  })
  .strict()
export type DeveloperToolModelRunResponseDto = z.infer<typeof developerToolModelRunResponseSchema>

export const developerToolHarnessProjectSchema = z
  .object({
    projectId: z.string().trim().min(1),
    displayName: z.string(),
    rootPath: z.string(),
  })
  .strict()
export type DeveloperToolHarnessProjectDto = z.infer<
  typeof developerToolHarnessProjectSchema
>
